### PR TITLE
Integrate `kissim_to_newick` tool into `kissim` code base

### PR DESCRIPTION
**Pull request by dominiquesydow**
## Description
Integrate `kissim_to_newick` tool into `kissim` code base and unify code style.

Functionality is accessible via (after pip installing the `kissim` package):
- Python API: 
  - `kissim.comparison.tree.from_file(inputfile, outputfile, clustering_method)`
  - `kissim.comparison.tree.from_distance_matrix(inputfile, outputfile, clustering_method)`
- CLI: `kissim tree -i inputfile -o outputfile -c method`

## Todos
- [x] Unify code style (snake case, f-strings, docstrings, ...)
- [x] Split `kissim_to_newick.py` into Python API and CLI parts
- [x] Rename `kissim_to_newick` to `tree`
- [x] Add problematic entries to top of the file (as global); if list grows think about moving it to a more prominent place accessible to all analyses functionalities
- [x] Add clustering method to signature (both for Python API and CLI)
- [x] Allow to generate tree from CSV file or matrix
- [x] Allow to save kinase annotation CSV file for FigTree annotation feature  
- [x] Add documentation for `tree` module
- [x] Replace `print` with `loggining`

## Questions
None

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by dominiquesydow** [comment]
Hi @AJK-dev, this is my suggestion for the tool integration into `kissim`. What do you think?
Please check the PR description for how the API and CLI signatures look like.

Supercedes https://github.com/volkamerlab/kissim/pull/57.

File: kissim/cli/main.py
Comment (AJK-dev) [inline kissim/cli/main.py:131]: Please add the default "ward" here, otherwise _None_ will be used as clustering method, which will result in an error.
`default="ward",`

File: kissim/comparison/tree.py
Comment (AJK-dev) [inline kissim/comparison/tree.py]: Small typo: problamatic vs  problematic.

File: kissim/comparison/tree.py
Comment (AJK-dev) [inline kissim/comparison/tree.py]: It would be a **nice-to-have** to make this an optional routine that can be disabled with an argument.
This because this routine takes longer than the tree creation routine and will in practice (for most people) only be relevant on the first run or when changing set.

File: kissim/comparison/tree.py
Comment (AJK-dev) [inline kissim/comparison/tree.py:199]: An alternative is to speed this call up by calling this function without any kinase names and afterwards filter the returned dataset.  Or cleaner: post multiple kinase names at once.
This will save you a few hundred API calls on every run -> a lot speedier ;-)

**Review by AJK-dev — CHANGES_REQUESTED** [review]
Very nice integration @dominiquesydow! Your coding skills are impressive and very nice to go through. I have a few suggestions for improvement.

File: kissim/comparison/tree.py
Comment (dominiquesydow) [reply kissim/comparison/tree.py:199]: Thank a lot for your quick and thorough review! 

I am unsure why I implemented this kinase query in `opencadd` per-kinase, does not make sense to me anymore. 
I updated `opencadd.databases.klifs` accordingly - while doing that I might have stumbled over a tiny bug in the KLIFS API, please check the PR https://github.com/volkamerlab/opencadd/pull/75

File: kissim/comparison/tree.py
Comment (dominiquesydow) [reply kissim/comparison/tree.py]: Thanks for catching this!

If the user does not want to write out the tree and annotation files, they can set `outputfile=None`. 
I had to move the `outputfile` variable within the functions, so that no errors are thrown if `None`.

Comment (dominiquesydow) [reply kissim/comparison/tree.py]: I just realized you were only commenting on the annotation not the writing-files procedure in general. 

I am in favor of letting the user set both output paths themselves:
`inputfile` > `matrix_path` (must be set)
`outputfile` > `tree_path` + `annotation_path` (can both be `None` and no output file will be written)

Comment (AJK-dev) [reply kissim/comparison/tree.py]: Even better!

**Comment by AJK-dev** [comment]
FYI I will do a last check and a few minor textual updates tomorrow.

**Review by AJK-dev — APPROVED** [review]
Everything seems to work nicely 👍. I'm approving now, but I won't yet merge it as you had one more item open (replacing the print).

**Comment by dominiquesydow** [comment]
Great, thank you, @AJK-dev! :)

Quick question: Why do you think the term "similarity" is more useful in this context than "distance"? 
AFAIK, similarity is often defined between 0 and 1 (the higher the more similar), while the distance is equal or greater than 0 (the lower the more similar). 

Since our matrix now shows distances, I thought we'd be safer to say "distances" instead of "similarity". 
Probably the word "distance" is too generic and that is why you changed it? If so, what do you think about "fingerprint distance" instead?

**Comment by AJK-dev** [comment]
> Quick question: Why do you think the term "similarity" is more useful in this context than "distance"?

My main goal with my latest commit was to at least ensure that the way it was used in the comments and the variable naming was at least correct. In several cases, the word _similarity_ was used where it should actually have been _distance_ and vice versa.
For the cases where I now used _similarity_, this is solely for the parts where the mean similarity is being used as the function actually calculates the mean similarity instead of the distance. 

That said, whether you choose to use distance or similarity is mostly a personal preference. However, I would argue that when you are generating a dendrogram you are looking for the similarities between the elements that are clustered together and not the dissimilarity. The goal, in the end, is (afaik) also to identify similarities, which is also implied by the name KISSIM 😉

> AFAIK, similarity is often defined between 0 and 1 (the higher the more similar), while the distance is equal or greater than 0 (the lower the more similar).

That depends on the similarity or distance measure that is being used. Distance is often also defined between 0 and 1 (the higher the more distant). I didn't look into it but assumed this was still the case. If the current distance measure is not ranging from 0-1 then the mean similarity calculation is wrong and should be corrected or simply adjusted to mean distance, depending on your preference and choice.
 
> Since our matrix now shows distances, I thought we'd be safer to say "distances" instead of "similarity".

The matrix is indeed now a distance matrix, but the current implementation for the calculation of the mean is the mean similarity, not the mean distance. 

> Probably the word "distance" is too generic and that is why you changed it? If so, what do you think about "fingerprint distance" instead?

This was not the reason as explained above. But please feel free to change the naming and mean calculation to whatever you like best.

**Comment by dominiquesydow** [comment]
Hi @AJK-dev, thanks a lot for clarifying this point. 
I indeed missed that we are converting the distance matrix into a similarity matrix to calculate a mean _similarity_ per node. Agreed, we stick to mean similarity (I added a flag that allows to also calculate mean distance).

File: kissim/comparison/tree.py
Comment (dominiquesydow) [inline kissim/comparison/tree.py]: Here we now convert the distance matrix into a similarity matrix (default) or not.

Since the `kissim` matrix now ranges from 0 to R+, I am normalizing before the `similarity = 1 - distance` step.

File: kissim/comparison/tree.py
Comment (dominiquesydow) [inline kissim/comparison/tree.py:239]: I also added here the root node mean similarity, this was a [fixed value before (1)](https://github.com/AJK-dev/kissim/blob/926aa5291b7f768f85c27bc6cb976c4db808401f/kissim/comparison/kissim_to_newick.py#L65). 

If this does not make any sense, I am happy to delete this again (for the newick format it does not make a change, since we do not write the root mean to file).

File: kissim/comparison/tree.py
Comment (dominiquesydow) [inline kissim/comparison/tree.py:273]: In case of the leaves, I also added for both the right and left leaf "1"/identity in case of similarity, this was set only for the [left leaf](https://github.com/AJK-dev/kissim/blob/926aa5291b7f768f85c27bc6cb976c4db808401f/kissim/comparison/kissim_to_newick.py#L86).

Again, happy to delete this if it does not make sense; no effect on the newick format as far as I understand.

File: kissim/comparison/tree.py
Comment (AJK-dev) [reply kissim/comparison/tree.py]: Seems like a good solution.
However, I would do the conversion to before calling this function. Since it's a recursive function, it's best to keep it as light as possible and to reduce the overhead.

One other thing to keep in mind: not having a fixed distance range will prevent the direct comparison of trees as in those cases the measure will depend on the input. This is at least the case for the current setup using the similarity. If the distance is not influenced by this, then it might be a reason to revert back and use the distance by default instead of the similarity.

File: kissim/comparison/tree.py
Comment (AJK-dev) [reply kissim/comparison/tree.py:273]: Good that you caught this. Indeed good to keep it this way as it might be useful in other circumstances/future formats.

File: kissim/comparison/tree.py
Comment (dominiquesydow) [reply kissim/comparison/tree.py]: > However, I would do the conversion to before calling this function. Since it's a recursive function, it's best to keep it as light as possible and to reduce the overhead.

Very good point, adapted this accordingly.

> One other thing to keep in mind: not having a fixed distance range will prevent the direct comparison of trees as in those cases the measure will depend on the input. This is at least the case for the current setup using the similarity. If the distance is not influenced by this, then it might be a reason to revert back and use the distance by default instead of the similarity.

Another good point - true, mean similarities could not be compared for different datasets. As you suggested I set the mean distance as default.

**Comment by dominiquesydow** [comment]
I think this PR is ready to go then :) Waiting for one more approval from you just in case.

**Review by AJK-dev — APPROVED** [review]
Go 😊!

**Pull request merged**
